iterators.tex

```diff
@@ -623,31 +650,186 @@
+\range{i}{j} denotes a range. Let $N$ be the smallest number of
+applications of \tcode{++i} necessary to make \tcode{bool(i == j)} be
+\tcode{true}. \tcode{SizedIterator<I>()} is satisfied if and only if:
+
+\begin{itemize}
+\item If $N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{j - i} is well-defined and equals $N$.
+
+\item If $-N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{i - j} is well-defined and equals $-N$.
+\end{itemize}
+\end{itemdescr}
+
+\pnum
+The \tcode{disable_sized} type predicate provides a mechanism to enable
+use of iterator types that meet the syntactic requirements but do not
+in fact satisfy \tcode{SizedIterator} with the library.
 
 \pnum
-The \tcode{Sentinel} concept defines requirements for a type that
+\enternote A program that instantiates a library template that requires
+an \tcode{Iterator} with such an iterator type \tcode{I} is ill-formed
+with no diagnostic required unless \tcode{disable_sized<I>} evalutes to
+\tcode{true}~(\ref{structure.requirements}). \exitnote
+
+\enternote The \tcode{SizedIterator} concept is satisfied by
+\tcode{RandomAccessIterator}s~(\ref{iterators.random.access}) and by
+counted iterators~(\ref{counted.iterator}).\exitnote
+
+\ednote{This concept also gives us a way to demote the category of \tcode{move_iterator}s to Input
+while retaining the ability of \tcode{move_iterator} pairs to communicate the range's size to
+container constructors.}
+} %% color{newclr}
+
+\rSec2[iterators.sentinel]{Concept Sentinel}
+\pnum
+The \tcode{Sentinel} concept \oldtxt{defines requirements for a type that
 is an abstraction of the past-the-end iterator. Its values can be
-compared to an iterator for equality.
+compared to an iterator for equality.}\newtxt{specifies requirements on
+the types \tcode{S} and \tcode{I} of a sentinel and an iterator that
+denote a range.}
 
 \indexlibrary{\idxcode{Sentinel}}%
-\begin{codeblock}
-  template <class T, class I>
+\begin{itemdecl}
+  template <class @\oldtxt{T}\newtxt{S}@, class I>
   concept bool Sentinel() {
-    return Regular<T>() &&
+    return Regular<@\oldtxt{T}\newtxt{S}@>() &&
       Iterator<I>() &&
-      EqualityComparable<T, I>();
+      EqualityComparable<@\oldtxt{T}\newtxt{S}@, I>();
   }
-\end{codeblock}
+\end{itemdecl}
+
+{\color{newclr}
+\begin{itemdescr}
+Let \tcode{s} and \tcode{t} be values of type \tcode{S},
+and \tcode{i} a value of type \tcode{I} such that \range{i}{s}
+denotes a range. Types \tcode{S} and \tcode{I} satisfy
+\tcode{Sentinel<S, I>()} if and only if:
+
+\begin{itemize}
+\item \tcode{i == s}, \tcode{s == i},\tcode{i != s}, and
+      \tcode{s != i} are well-defined.
+
+\item \tcode{bool(i == s)} if and only if \tcode{i} and
+      \tcode{s} denote the same element.
+
+\item If \range{i}{s} and \range{i}{t} denote the same range,
+      then \tcode{s == t} is well-defined and \tcode{s}
+      equals \tcode{t}.
```

Comment:
> I haven't yet tried to specify precisely the domain of == for sentinels.
I haven't bothered since == on sentinels is a pretty useless operation, but
you've convinced me that I need to look closer.

> + item \tcode{bool(i == s)} if and only if \tcode{i} and
> +      \tcode{s} denote the same element.

Just reading through a recent draft. Sentinels are defined as abstractions
of PTE iterators, so "denote the same element" seems weird.


> +\item If \range{i}{s} and \range{i}{t} denote the same range,
> +      then \tcode{s == t} is well-defined and \tcode{s}
> +      equals \tcode{t}.

My understanding is that sentinels are essentially predicates. So two
functions (expressions, whatever) are equal iff they map the same inputs to
the same outputs. That's tough to implement in C++ :) But you don't really
need to. Representation is a sufficient approximation (function pointer,
function object type, etc).

But this axiom isn't actually true. s and t could be different functions
whose mappings overlap for that particular sequence of elements.

I suppose you could say that two sentinels s and t are equal if, for all
iterators i, i==s equals i==t. Then it would be the case that if s and t
are the same predicate then [i, s) and [i, t) would be the same ranges.

My 2c.